---
title: "Saez Lab tools on transcriptomics from intestinal organoids treated with Sars-CoV-2"
author: "Alberto Valdeolivas: alberto.valdeolivas@bioquant.uni-heidelberg.de; Date:"
date: "14/07/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### License Info

This program is free software: you can redistribute it and/or modify it under 
the terms of the GNU General Public License as published by the Free Software 
Foundation, either version 3 of the License, or (at your option) any later 
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR 
A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Please check http://www.gnu.org/licenses/.

## Introduction

The present script takes the results of the differential expression analysis on
transcriptomics data from intestinal organoids treated with Sars-CoV-2 for 60 
hours and the following tools: Progeny, Dorothea and CARNIVAL (see References). 

The original transcriptomic dataset is coming from the following publication:
[Lamers et al. 2020](https://science.sciencemag.org/content/369/6499/50).

The differential expression analysis was conducted by Martina Poletti (<Martina.Poletti@earlham.ac.uk>) 

### Getting Ready 

We first load the required libraries and we define a function to export CARNIVAL
results to Cytoscpae. 

```{r, message=FALSE}
library(tibble)
library(dplyr)
library(readr)
library(progeny)
library(dorothea)
library(CARNIVAL)
library(ggplot2)
library(OmnipathR)
library(biomaRt)
library(grid)
library(gridExtra)

## We also define a function to format the CARNIVAL output to cytoscape
OutputCyto <- function(CarnivalResults, outputFile) {
    CarnivalNetwork <- 
        as.data.frame(CarnivalResults$weightedSIF, stringsAsFactors = FALSE) %>%
        dplyr::mutate(Sign = as.numeric(Sign), Weight = as.numeric(Weight)) %>% 
        dplyr::mutate(Weight = Sign * Weight) %>%
        dplyr::select(Node1, Weight, Node2)
        
    CarnivalNetworkNodes <- 
        unique(c(CarnivalNetwork$Node1,CarnivalNetwork$Node2))
    
    CarnivalAttributes <- CarnivalResults$nodesAttributes %>% 
        as.data.frame() %>%
        dplyr::filter(Node %in% CarnivalNetworkNodes) %>%
        dplyr::mutate(NodeType = as.character(NodeType)) %>%
        dplyr::mutate(NodeType=if_else(NodeType =="", "I", NodeType))
            
    nameOutputNetwork <- paste0(outputFile, "Network.sif")
    nameOutputAttributes <-  paste0(outputFile, "Attributes.txt")    
    
    write.table(CarnivalNetwork, file = nameOutputNetwork,
        quote = FALSE, row.names = FALSE, col.names = FALSE, sep = " ")
    
    write.table(CarnivalAttributes, file = nameOutputAttributes,
        quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
}
```

We also read the results from the differential expression analysis: 

```{r, message=FALSE}
## There are two duplicates Mar-02 and RPL21. I am going to remove them
## in a first approach
all_genes_dea <- read_csv("InputData/Clevers_degs_all.csv") %>% 
    dplyr::filter(!(Gene %in% c("Mar-02","RPL21")))
```

## Results 

### Pathway activity estimation using Progeny

We first estimate the pathway activity using the Progeny package. In particular,
we compute the normalised enriched score (NES) of the different pathways by 
running Progeny using the statistic from the differential express analysis.

```{r, message=FALSE}
all_genes_dea_stat <- all_genes_dea %>% 
    dplyr::select(Gene, stat) %>% 
    dplyr::filter(!is.na(stat)) %>% 
    column_to_rownames(var = "Gene") 

pathways_zscore <- t(progeny(as.matrix(all_genes_dea_stat), 
    scale=TRUE, organism="Human", top = 100, perm = 10000, z_scores = TRUE))
colnames(pathways_zscore) <- "NES"

## I also need to run progeny in such a way to have values between 1 and -1 to
## use as CARNIVAL input
pathways_inputCarnival <- 
  t(progeny(as.matrix(all_genes_dea_stat), 
    scale=TRUE, organism="Human", top = 100, perm = 10000, z_scores = FALSE))
colnames(pathways_inputCarnival) <- "Activity"
```

We now display the normalized enrichment scores (NES) in a bar plot.

```{r, dpi=300}
pathways_zscore_df <- as.data.frame(pathways_zscore) %>% 
    rownames_to_column(var = "Pathway") %>%
    dplyr::arrange(NES) %>%
    dplyr::mutate(Pathway = factor(Pathway))

ggplot(pathways_zscore_df,aes(x = reorder(Pathway, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Pathways")
```

It is quite shocking that the MAPK pathway in un-active. This pathway was 
usually found to be more active after SARS-CoV-2 infection in other studies.

```{r, warning=FALSE, message=FALSE}
weight_matrix <- getModel("Human", top=100)
weight_matrix <- data.frame(names = row.names(weight_matrix), 
  row.names = NULL, weight_matrix)
plots <- progenyScatter(all_genes_dea_stat %>% 
  tibble::rownames_to_column(var = "GeneID"), weight_matrix)
```

We therefore check which genes are contributing the most to the MAPK progeny 
score. In these plots, we have the progeny scores for each pathway and stats
of the different genes. In red, we will have the genes contributing to the 
positive activation score and on blue the opposite. 

```{r, dpi=300}
grid.draw(plots[[1]]$MAPK)
```

We also check for the JAK-STAT pathway:

```{r, dpi=300}
grid.draw(plots[[1]]$JAK.STAT)
```

### Transcription factor activity with Dorothea and Viper. 

Now, we estimate the transcription factor (TF) activity using the dorothea 
package. We select Dorothea interactions with confidence level A, B and C.

```{r}
## We load Dorothea Regulons
data(dorothea_hs, package = "dorothea")
regulons <- dorothea_hs %>%
  dplyr::filter(confidence %in% c("A", "B","C"))
```

We run Viper using the statistic from the different expression analysis. First,
we run it considering TF with at least 5 targets and with no correction for 
pleiotropic regulation

```{r, message=FALSE}
all_genes_dea_stat <-  all_genes_dea %>% 
    dplyr::select(Gene, stat) %>% 
    dplyr::filter(!is.na(stat)) %>% 
    column_to_rownames(var = "Gene") %>%
    as.matrix()

tf_activities_stat <- 
    dorothea::run_viper(all_genes_dea_stat, regulons,
    options =  list(minsize = 5, eset.filter = FALSE, 
    cores = 8, verbose = FALSE, nes = TRUE))
```

We now display the top 25 normalized enrichment scores (NES) for the TF in a 
bar plot.

```{r, dpi=300}
tf_activities_top25 <- tf_activities_stat %>%
    as.data.frame() %>% 
    rownames_to_column(var = "Tf") %>%
    dplyr::rename(NES = "stat") %>%
    dplyr::top_n(25, wt = abs(NES)) %>%
    dplyr::arrange(NES) %>% 
    dplyr::mutate(Tf = factor(Tf))

ggplot(tf_activities_top25,aes(x = reorder(Tf, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Transcription Factors")
```

I run again Viper with a more conservative setup. The TFs need to regulate at 
least 15 targets genes and I include the correction for pleiotropic regulation 

```{r, dpi=300}
tf_activities_stat_pleitropic <- 
    dorothea::run_viper(all_genes_dea_stat, regulons,
    options =  list(minsize = 15, eset.filter = FALSE, 
    cores = 8, verbose = FALSE, nes = TRUE, pleiotropy= TRUE))

tf_activities_top25_pleitropic <- tf_activities_stat_pleitropic %>%
    as.data.frame() %>% 
    rownames_to_column(var = "Tf") %>%
    dplyr::rename(NES = "stat") %>%
    dplyr::top_n(25, wt = abs(NES)) %>%
    dplyr::arrange(NES) %>% 
    dplyr::mutate(Tf = factor(Tf))

ggplot(tf_activities_top25_pleitropic,aes(x = reorder(Tf, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Transcription Factors")
```


### Running CARNIVAL

We use the OmnipathR package to fetch the Omnipath database and generate the 
prior knowledge network. We take the signed and directed protein-protein 
interactions. 

```{r}
ia_omnipath <- import_omnipath_interactions() %>% as_tibble()
ia_pwextra <- import_pathwayextra_interactions() %>% as_tibble()
ia_kinaseextra <- import_kinaseextra_interactions() %>% as_tibble()

## We bind the datasets
interactions <- as_tibble(
    bind_rows(
        ia_omnipath %>% mutate(type = 'ppi'),
        ia_pwextra %>% mutate(type = 'ppi'),
        ia_kinaseextra %>% mutate(type = 'ppi')))

signed_directed_interactions <- 
    dplyr::filter(interactions, consensus_direction==1) %>%
    filter(consensus_stimulation == 1 | consensus_inhibition == 1) %>% 
    dplyr::mutate(sign = if_else(consensus_stimulation==1,1,-1))  %>%
    dplyr::select(source_genesymbol, sign,  target_genesymbol) %>%
    dplyr::rename(source ="source_genesymbol", target ="target_genesymbol")

carnival_pkn <- signed_directed_interactions %>%
    dplyr::distinct(source, target, .keep_all = TRUE)

all_source_nodes <- unique(carnival_pkn$source)
all_target_nodes <- unique(carnival_pkn$target)
all_nodes_network <- unique(c(all_source_nodes,all_target_nodes))
```

We are going to explore the host-virus interactions from IntAct. 

```{r}
host_viral_interactions <- 
  read_tsv(file = "InputData/sarscov2-human_viral_ppis_Intact_June2020.txt")

## We translate the human uniprot symbols to hgnc. 
ensembl <- useMart('ensembl', dataset="hsapiens_gene_ensembl")
# listAttributes(ensembl)
uniprot_hgnc <- 
  getBM(attributes=c("uniprotswissprot", "hgnc_symbol"),  
      filters="uniprotswissprot",
      values=unique(host_viral_interactions$human_protein), mart=ensembl)

host_viral_interactions_hgnc <- 
  dplyr::inner_join(host_viral_interactions, uniprot_hgnc, 
                    by = c("human_protein" = "uniprotswissprot"))

## These are the number of interactions: 
nrow(host_viral_interactions_hgnc)

## Are these human proteins sources of any interaction in our prior knowledge
## network? Otherwise it does not make sense to include them as perturbartions 
## of the signaling. We assume that they are inhibitions hampering the regular
## signaling 
host_viral_interactions_hgnc_filter <- host_viral_interactions_hgnc %>% 
  dplyr::filter(hgnc_symbol %in% all_source_nodes) %>% 
  dplyr::select(viral_protein, hgnc_symbol) %>% 
  dplyr::distinct() %>% 
  dplyr::mutate(sign = -1) %>% 
  dplyr::rename(source = "viral_protein", target = "hgnc_symbol") %>%
  dplyr::select(source, sign, target)
```

We finally merge the virus-host interactions with our prior knowledge network
and we define the viral proteins as perturbations. We also prepare TF activity
scores and progeny weights. In particular, we are going to select the top 50 TFs
when running viper with the pleitropic correction and the TFs should have at 
least 15 targets. 

```{r}
## Prior Knowledge Network
carnival_pkn_hostvirus <- 
  bind_rows(carnival_pkn, host_viral_interactions_hgnc_filter)

## Perturbation (Viral proteins are active)
viral_proteins <- unique(host_viral_interactions_hgnc_filter$source)
viral_proteins_perturbation <- 
  data.frame(viral_proteins = viral_proteins, sign = 1) %>% 
  tibble::column_to_rownames(var = "viral_proteins") %>%
  t() %>% as.data.frame()

## Top TFs
tf_top50_pleitropic <- tf_activities_stat_pleitropic %>%
  as.data.frame() %>% 
  rownames_to_column(var = "TF") %>%
  dplyr::filter(TF %in% all_nodes_network) %>% 
  dplyr::rename(NES = "stat") %>%
  dplyr::top_n(50, wt = abs(NES)) %>%
  dplyr::arrange(NES) %>%
  tibble::column_to_rownames(var = "TF")  %>%
  t() %>% as.data.frame()

## Progeny Weigths
progeny_weigths <- pathways_inputCarnival %>% t() 
```

And we run CARNIVAL: 

```{r, message=FALSE, warning=FALSE, eval=FALSE}
carnival_results_top50tf_pleitropic_minsize15 <-runCARNIVAL(
    solverPath="/opt/ibm/ILOG/CPLEX_Studio129/cplex/bin/x86-64_linux/cplex",
    netObj=carnival_pkn_hostvirus,
    measObj=tf_top50_pleitropic,
    inputObj = viral_proteins_perturbation,
    # dir_name="Carnival_Results",
    weightObj=progeny_weigths,
    # nodeID = 'gene',
    timelimit = 7200,
    solver = "cplex")
saveRDS(carnival_results_top50tf_pleitropic_minsize15, 
    file = "Carnival_Results/carnival_results_top50tf_pleitropic_minsize15.rds")
OutputCyto(carnival_results_top50tf_pleitropic_minsize15, 
    outputFile="Carnival_Results/carnival_results_top50tf_pleitropic_minsize15")
```

Network with the results: Rectangles are the most active transcription factors 
after infection and the inverse triangles are the perturbed nodes. Ellipses 
are signaling intermediates proteins linking those perturbations and TFs. 
Red means activation after infection and blue the opposite. 

<br><br>
![](Carnival_Results/carnival_results_top50tf_pleitropic_minsize15Network.sif.png)
<br><br>

I am also going to run CARNIVAL by assuming that the virus-host protein 
interactions are activatory. 

```{r, message=FALSE, warning=FALSE, eval=FALSE}
host_viral_interactions_hgnc_filter_viralActivation <- 
  host_viral_interactions_hgnc_filter %>%
  dplyr::mutate(sign = 1) 

## Prior Knowledge Network
carnival_pkn_hostvirus <- 
  bind_rows(carnival_pkn, host_viral_interactions_hgnc_filter_viralActivation)

carnival_results_top50tf_pleitropic_minsize15_activation <-runCARNIVAL(
    solverPath="/opt/ibm/ILOG/CPLEX_Studio129/cplex/bin/x86-64_linux/cplex",
    netObj=carnival_pkn_hostvirus,
    measObj=tf_top50_pleitropic,
    inputObj = viral_proteins_perturbation,
    # dir_name="Carnival_Results",
    weightObj=progeny_weigths,
    # nodeID = 'gene',
    timelimit = 7200,
    solver = "cplex")
saveRDS(carnival_results_top50tf_pleitropic_minsize15_activation, 
    file = "Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_activation.rds")
OutputCyto(carnival_results_top50tf_pleitropic_minsize15_activation, 
    outputFile="Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_activation")
```
<br><br>
![](Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_activationNetwork.sif.png)
<br><br>

Finally, I am going to run CARNIVAL defining the perturbations (human proteins
that interact with the viral proteins) but without defining their effect 
(stimulation or inhibition). CARNIVAL will infer the effect. 

```{r, message=FALSE, warning=FALSE, eval=FALSE}
human_proteins_undefined_perturbation <- 
  host_viral_interactions_hgnc_filter %>% 
  dplyr::mutate(sign = NaN) %>%
  dplyr::select(target, sign) %>% 
  dplyr::mutate(sign = as.numeric(sign)) %>% 
  dplyr::distinct() %>% 
  dplyr::filter(target %in% all_source_nodes) %>%
  tibble::column_to_rownames(var = "target") %>% 
  t() %>% as.data.frame()

carnival_results_top50tf_pleitropic_minsize15_undefinedEffect <-runCARNIVAL(
    solverPath="/opt/ibm/ILOG/CPLEX_Studio129/cplex/bin/x86-64_linux/cplex",
    netObj=carnival_pkn,
    measObj=tf_top50_pleitropic,
    inputObj = human_proteins_undefined_perturbation,
    # dir_name="Carnival_Results",
    weightObj=progeny_weigths,
    # nodeID = 'gene',
    timelimit = 14400,
    solver = "cplex")
saveRDS(carnival_results_top50tf_pleitropic_minsize15_undefinedEffect, 
    file = "Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_undefinedEffect.rds")
OutputCyto(carnival_results_top50tf_pleitropic_minsize15_undefinedEffect, 
    outputFile="Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_undefinedEffect")
```

<br><br>
![](Carnival_Results/carnival_results_top50tf_pleitropic_minsize15_undefinedEffectNetwork.sif.png)
<br><br>

## Important References 

**Progeny** publication: 

> Schubert M, Klinger B, Klünemann M, Sieber A, Uhlitz F, Sauer S, Garnett MJ, Blüthgen N, Saez-Rodriguez J. “Perturbation-response genes reveal signaling footprints in cancer gene expression.” _Nature Communications_: [10.1038/s41467-017-02391-6](https://doi.org/10.1038/s41467-017-02391-6)

> Holland CH, Szalai B, Saez-Rodriguez J. "Transfer of regulatory knowledge from human to mouse for functional genomics analysis." _Biochimica et Biophysica Acta (BBA) - Gene Regulatory Mechanisms._ 2019. DOI: [10.1016/j.bbagrm.2019.194431](https://doi.org/10.1016/j.bbagrm.2019.194431).

**Dorothea** publication:

> Garcia-Alonso L, Holland CH, Ibrahim MM, Turei D, Saez-Rodriguez J. "Benchmark and integration of resources for the estimation of human transcription factor activities." _Genome Research._ 2019. DOI: [10.1101/gr.240663.118](https://doi.org/10.1101/gr.240663.118).

**CARNIVAL** publication:

> Liu A., Trairatphisan P., Gjerga E. et al. [From expression footprints to causal pathways: contextualizing large signaling networks with CARNIVAL](https://www.nature.com/articles/s41540-019-0118-z) _npj Systems Biology and Applications_ volume 5, Article number: 40 (2019) (equal contributions).

**Organoid Transcriptomics data** related publication:

> MART M. LAMERS, JOEP BEUMER, JELTE VAN DER VAART, KÈVIN KNOOPS, JENS PUSCHHOF, TIM I. BREUGEM, RAIMOND B. G. RAVELLI, J. PAUL VAN SCHAYCK, ANNA Z. MYKYTYN, HANS Q. DUIMEL, ELLY VAN DONSELAAR, SAMRA RIESEBOSCH, HELMA J. H. KUIJPERS, DEBBY SCHIPPER, WILLINE J. VAN DE WETERING, MIRANDA DE GRAAF, MARION KOOPMANS, EDWIN CUPPEN, PETER J. PETERS, BART L. HAAGMANS, HANS CLEVERS [SARS-CoV-2 productively infects human gut enterocytes](https://science.sciencemag.org/content/369/6499/50), _Science_ July 202. 


## Session Info Details

```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```
